File: kotlinRpnSolver/src/parse/ShuntingYardParser.kt
package parse

import java.util.Stack
import kotlin.math.pow

import exceptions.AggregateException
import token.*
import java.util.ArrayList

fun tokensToRpn(tokens: List<Token>): List<Token> {
    tokens.throwIfContainsInvalid()

    val operatorStack = Stack<OperatorToken>()
    val rpnSolveList = ArrayList<Token>()

    for (token in tokens) {
        when (token) {
            is LiteralToken -> rpnSolveList.add(token)
            is OperatorToken -> {
                when (token.operator) {
                    Operator.RPAREN -> {
                        while (true) {
                            if (operatorStack.isEmpty()) {
                                throw IllegalArgumentException("Mismatched Parentheses in token list\n\nSolve List:${rpnSolveList}\n\nOperator Stack:${operatorStack}")
                            }
                            if (!operatorStack.pop().run {
                                    when (operator) {
                                        Operator.LPAREN -> false
                                        else -> rpnSolveList.add(this)
                                    }
                                } || operatorStack.isEmpty()) {
                                break;
                            }
                        }
                    }

                    else -> {
                        while (operatorStack.isNotEmpty() && precedenceCompare(token, operatorStack.peek())) {
                            rpnSolveList.add(operatorStack.pop())
                        }
                        operatorStack.push(token)
                    }
                }
            }
        }
    }
    while (operatorStack.isNotEmpty()) {
        rpnSolveList.add(operatorStack.pop())
    }
    return rpnSolveList
}

private fun precedenceCompare(leftOpToken: OperatorToken, rightOpToken: OperatorToken): Boolean = when (rightOpToken.associativity) {
    Associativity.LEFT -> leftOpToken.precedence() <= rightOpToken.precedence()
    Associativity.RIGHT -> leftOpToken.precedence() < rightOpToken.precedence()
    Associativity.NONE -> false
}


fun shuntingYardSolve(rpnSolveList: List<Token>): Pair<Double, List<String>> {
    rpnSolveList.throwIfContainsInvalid()

    val solveStack = Stack<Token>()
    val steps = ArrayList<String>()
    steps.add("Showing steps for RPN solve:")
    var stepCount = 0
    var stepChanged = true

    for (token in rpnSolveList) {
        when (token) {
            is LiteralToken -> {
                solveStack.push(token)
            }

            is OperatorToken -> {
                when (token.numOperands) {
                    2 -> {
                        val lhs = solveStack.pop()
                        val rhs = solveStack.pop()
                        if (lhs !is LiteralToken || rhs !is LiteralToken) {
                            throw IllegalArgumentException("Operator arguments were not both literals. lhs: ${lhs.toSimpleString()} rhs: ${rhs.toSimpleString()}.")
                        }
                        val arithmeticResult = when (token.operator) {
                            Operator.PLUS -> lhs.value + rhs.value
                            Operator.MINUS -> lhs.value - rhs.value
                            Operator.MULTIPLY -> lhs.value * rhs.value
                            Operator.DIVIDE -> lhs.value / rhs.value
                            Operator.EXPONENT -> lhs.value.pow(rhs.value)
                            else -> throw IllegalArgumentException("Invalid operator provided: ${token.toSimpleString()}, number of operands: ${token.numOperands}. Binary operator expected.")
                        }
                        stepCount++;
                        stepChanged = true;
                        steps.add("BINARY OPERATION ${token.toSimpleString()} ON ${lhs.toSimpleString()} AND ${rhs.toSimpleString()} => INSERT {$arithmeticResult}");

                        solveStack.push(LiteralToken(arithmeticResult))
                    }

                    1 -> {
                        val (arithmeticResult, operand) = solveStack.pop().let { operand ->
                            when (operand) {
                                is LiteralToken -> when (token.operator) {
                                    Operator.UNARY_PLUS -> Pair(+operand.value, operand)
                                    Operator.UNARY_MINUS -> Pair(-operand.value, operand)
                                    else -> throw IllegalArgumentException("Invalid operator provided: ${token.toSimpleString()}, number of operands: ${token.numOperands}. Unary operator expected.")
                                }

                                else -> throw IllegalArgumentException("Operator argument was not a literal. lhs: ${operand.toSimpleString()}.")
                            }
                        }
                        stepCount++;
                        stepChanged = true;
                        steps.add("UNARY OPERATION ${token.toSimpleString()} ON ${operand.toSimpleString()} => INSERT {$arithmeticResult}");

                        solveStack.push(LiteralToken(arithmeticResult))
                    }

                    else -> throw IllegalArgumentException("Invalid number of operands ${token.numOperands} for operator ${token.operator}")
                }
            }
        }
    }
    return solveStack.singleOrNull()?.let {
        when (it) {
            is LiteralToken -> Pair(it.value, steps)
            else -> throw IllegalStateException("Final token is not a LiteralToken")
        }
    } ?: throw IllegalArgumentException("Final result was not a single value. Result stack count: ${solveStack.count()}, Result stack values: ${solveStack.joinToString(",") { it.toSimpleString() }}")
}

private fun List<Token>.throwIfContainsInvalid() {
    filterIsInstance<InvalidToken>().run {
        if (isNotEmpty())
            throw AggregateException("Parsing errors encountered:", map { IllegalArgumentException("Invalid token: ${it.invalidShard} at position ${it.position}") })
    }
}

File: CSharpRPNSolver/ExpressionParser/Parse/ShuntingYardParser.cs
using ExpressionParser.Tokens;
using System.Text.RegularExpressions.Generated;

namespace ExpressionParser.Parse;

internal static class ShuntingYardParser
{
    public static decimal Parse(List<Token> tokens)
    {
        var rpnList = BuildRPNFrom(tokens);
        return SolveRPNOf(rpnList).result;
    }

    public static List<Token> BuildRPNFrom(List<Token> tokens)
    {
        IEnumerable<InvalidToken> invalidTokens = tokens.OfType<InvalidToken>();
        if (invalidTokens.Any())
        {
            IEnumerable<ArgumentException> invalidTokenExceptions = invalidTokens.Select(t => new ArgumentException($"Invalid token: {t.InvalidShard} at position {t.Index}"));
            throw new AggregateException("Parsing errors encountered:", invalidTokenExceptions);
        }

        var operatorStack = new Stack<OperatorToken>();
        var rpnSolveList = new List<Token>();

        // build the solve list
        foreach (var token in tokens)
        {
            if (token is LiteralToken)
            {
                rpnSolveList.Add(token);
            }
            else if (token is OperatorToken opToken)
            {
                if (opToken.Value == Operator.RPAREN)
                {
                    do
                    {
                        if (operatorStack.Count == 0)
                        {
                            throw new InvalidOperationException($"{{ShuntingYardParser:Parse}}: Right Paren was missing a corresponding Left Paren");
                        }
                        var op = operatorStack.Pop();
                        if (op.Value != Operator.LPAREN)
                        {
                            rpnSolveList.Add(op);
                        }
                        else
                        {
                            break;
                        }
                    } while (operatorStack.Count > 0);
                }
                else
                {
                    // check precedence
                    bool PrecedenceCompare(OperatorToken otherToken)
                    {
                        if (otherToken.Associativity == Associativity.LEFT)
                        {
                            return opToken.Precedence <= otherToken.Precedence;
                        }
                        else if (otherToken.Associativity == Associativity.RIGHT)
                        {
                            return opToken.Precedence < otherToken.Precedence;
                        }
                        else //Associativity.NONE (paren)
                        {
                            return false;
                        }
                    }
                    while (operatorStack.Count > 0 && PrecedenceCompare(operatorStack.Peek()))
                    {
                        rpnSolveList.Add(operatorStack.Pop());
                    }
                    operatorStack.Push(opToken);
                }
            }
        }

        // dump remaining operators from operator stack
        while (operatorStack.Count > 0)
        {
            rpnSolveList.Add(operatorStack.Pop());
        }

        return rpnSolveList;
    }

    public static (decimal result, List<string> steps) SolveRPNOf(List<Token> rpnSolveList)
    {
        IEnumerable<InvalidToken> invalidTokens = rpnSolveList.OfType<InvalidToken>();
        if (invalidTokens.Any())
        {
            IEnumerable<ArgumentException> invalidTokenExceptions = invalidTokens.Select(t => new ArgumentException($"Invalid token: {t.InvalidShard} at position {t.Index}"));
            throw new AggregateException("Parsing errors encountered:", invalidTokenExceptions);
        }
        // resolve the solve list
        List<string> steps = ["Showing steps for RPN solve:"];

        var step = 0;
        var stepChanged = true;
        for (int i = 0; i < rpnSolveList.Count && rpnSolveList.Count > 1; i++)
        {
            if (rpnSolveList[i] is OperatorToken opToken)
            {
                if (i < opToken.NumOperands)
                {
                    throw new InvalidOperationException($"{{ShuntingYardParser:Parse}}: Operator too far left: missing arguments");
                }
                else
                {
                    if (opToken.NumOperands == 2)
                    {
                        if (rpnSolveList[i - 2] is LiteralToken lhs && rpnSolveList[i - 1] is LiteralToken rhs)
                        {
                            var arithmeticResult = opToken.Value switch
                            {
                                Operator.MULTIPLY => lhs.Value * rhs.Value,
                                Operator.DIVIDE => lhs.Value / rhs.Value,
                                Operator.PLUS => lhs.Value + rhs.Value,
                                Operator.MINUS => lhs.Value - rhs.Value,
                                Operator.EXPONENT => (decimal)Math.Pow((double)lhs.Value, (double)rhs.Value)
                            };

                            step++;
                            stepChanged = true;
                            steps.Add($"BINARY OPERATION {opToken.ToSimpleString()} ON {lhs.ToSimpleString()} AND {rhs.ToSimpleString()} => INSERT {{{arithmeticResult}}}");

                            rpnSolveList.RemoveAt(i);
                            rpnSolveList.RemoveAt(i - 1);
                            rpnSolveList[i - 2] = new LiteralToken(arithmeticResult);
                            i -= 2;
                        }
                        else
                        {
                            throw new InvalidOperationException($"{{ShuntingYardParser:Parse}}: Operator arguments were not both literals");
                        }
                    }
                    else if (opToken.NumOperands == 1)
                    {
                        if (rpnSolveList[i - 1] is LiteralToken operand)
                        {
                            var arithmeticResult = opToken.Value switch
                            {
                                Operator.UNARY_PLUS => +operand.Value,
                                Operator.UNARY_MINUS => -operand.Value,
                            };

                            step++;
                            stepChanged = true;
                            steps.Add($"UNARY OPERATION {opToken.ToSimpleString()} ON {operand.ToSimpleString()} => INSERT {{{arithmeticResult}}}");

                            rpnSolveList.RemoveAt(i);
                            rpnSolveList[i - 1] = new LiteralToken(arithmeticResult);
                            i--;
                        }
                        else
                        {
                            throw new InvalidOperationException($"{{ShuntingYardParser:Parse}}: Operator argument for op: {opToken.Value} was not a literal");
                        }
                    }
                    else
                    {
                        throw new InvalidOperationException($"{{ShuntingYardParser:Parse}}: Invalid number of operands {opToken.NumOperands} for operator {opToken.Value}");
                    }
                }
            }

            if (stepChanged)
            {
                steps.Add($"step {step} list: {string.Join(' ', rpnSolveList.Select(token => token.ToSimpleString()))}");
                stepChanged = false;
            }
        }

        if (rpnSolveList.Count == 1 && rpnSolveList[0] is LiteralToken litResult)
        {
            return (litResult.Value, steps);
        }
        throw new InvalidOperationException($"{rpnSolveList}");
    }
}
